### PR TITLE
murdock: also compile with LLVM/clang

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This is just an RFC and just a sketch at the moment, as I don't know how good this idea is (may slow down Murdock builds, but so would any new board) or if what I'm doing here is even correct.

But there are two major reasons for this:

1. clang picks up different errors sometimes than GCC.
2. OSX support is hardened as it is usually the toolchain used there.

### Issues/PRs references
None (but idea sparked by https://github.com/RIOT-OS/RIOT/pull/8646#discussion_r197372867)

### CODE REVIEW

**Comment by miri64** [comment]
Requested review by @kaspar030 since it is his code I'm fiddling around with and by @gebart since IIRC he introduced the `TOOLCHAIN` feature.

**Comment by miri64** [comment]
Seems to work now + Murdock already picked out some errors ;-)

File: .murdock [hunk 1]
```diff
@@ -76,21 +76,34 @@ get_supported_boards() {
...
-# board. Only print for boards in $BOARDS.
-get_app_board_pairs() {
+get_supported_toolchains() {
+    local board=$1
+
+    if [ "$board" = "native" ]; then
+        echo "gnu llvm"
+    else
+        echo "gnu"
+    fi
```
Comment (miri64) [inline .murdock]: Maybe this construct could be replaced with some `BOARD=$board make info-toolchains-supported` target? This way also the ARM platforms will be build with LLVM.

**Comment by miri64** [comment]
> Seems to work now + Murdock already picked out some errors ;-)

Except for [one error in `can`](https://ci.riot-os.org/RIOT-OS/RIOT/9398/140456b7d689b1ec4d39c58ad64178230795514d/output/compile/tests/conn_can/native/llvm.txt) all seem to be in packets...

**Comment by vincent-d** [comment]
> Except for one error in can all seem to be in pkgs...

Funny one, this seems to come from `LL_PREPEND_ELEM` from `utlist.h`. Not a clue how to fix it though!

**Comment by gebart** [comment]
I think the best way is just to ignore the Wtautological-pointer-compare warning for that line.
See http://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas

Use something in utlist.h like the example in the clang manual

```c
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wtautological-pointer-compare"

Blah

#pragma clang diagnostic pop
```

**Comment by cladmi** [comment]
I would really like to have llvm for native and llvm for at least few reference boards.
Just doing one board per architecture could already help maintaining the build system.

**Comment by miri64** [comment]
@cladmi any idea how to achieve this based on the changes I already provided?

**Comment by cladmi** [comment]
@miri64 I would say it depends, on what is wanted, test all possible boards or just some.

If it is focusing first only on CI where some boards are selected for testing, adding other names at the place where you put `native` and check if the board is in the "test with llvm" list. Using https://stackoverflow.com/a/24495744/395687 oneliner for example.
Good thing, helps starting

If it is meant to be global, the `info-toolchains-supported` or maybe only `info-toolchains-supported-on-ci` could be an idea too as the configuration could be put in cpu common directories. But for the global `info-toolchains-supported`, as a user I think I would expect to check the computer configuration, like is llvm installed, does it support this architecture, which is a feature on its own and maybe not only read a static configuration that could fail for me.

**Comment by miri64** [comment]
> Just doing one board per architecture could already help maintaining the build system.

> If it is focusing first only on CI where some boards are selected for testing, adding other names at the place where you put native and check if the board is in the "test with llvm" list. Using https://stackoverflow.com/a/24495744/395687 oneliner for example.
Good thing, it starts small and does not double the number of jobs.

> […] which is a feature on its own and maybe not only read a static configuration that could fail for me.

Ok, then considering these three statements and with the urge to keep it simple for now, I would then check platform is native or a selection of a few other ARM-based boards in the hard-coded manner I applied for now and we later on select to provide a make target.

**Comment by miri64** [comment]
> or a selection of a few other ARM-based boards

(to my knowledge non of our other architectures have llvm-support at the moment)

**Comment by miri64** [comment]
I added `samr21-xpro` and one random nucleo board with Cortex-M4 now. Most other boards I tried did not compile :(

**Comment by cladmi** [comment]
It looks like murdock cannot generated the aggregated report with this added parameter.

From guessing and if I understand correctly I would imagine that adding `-E TOOLCHAIN` to `DWQ_ENV` may help there.
If this is used to separate jobs, I think it would prevent two different build to generate an output. @kaspar030 would know.

**Comment by miri64** [comment]
Let's try that ;-)

File: .murdock [hunk 2]
```diff
@@ -76,21 +76,38 @@ get_supported_boards() {
...
 }
 
-# given an app dir as parameter, print "$appdir board" for each supported
-# board. Only print for boards in $BOARDS.
-get_app_board_pairs() {
+get_supported_toolchains() {
+    local board=$1
+
+    # TODO provide make target in similar fashion as the info-boards-supported one
+    for b in native samr21-xpro nucleo-f401re
```
Comment (cladmi) [inline .murdock]: I would prefer to have this list in a global variable outside of the function to not have this "magic" list in the function.

**Comment by cladmi** [comment]
@kaspar030 do you have any idea how to fix the output ?

File: .murdock (hunk 2 above)
Comment (kaspar030) [reply .murdock]: +1, similar to TEST_BOARDS_AVAILABLE

Comment (kaspar030) [inline .murdock]: please re-use the ```is_in_list``` function.

**Comment by kaspar030** [comment]
> do you have any idea how to fix the output ?

I'll have a look.

File: .murdock (hunk 2 above)
Comment (miri64) [reply .murdock]: Ok

Comment (miri64) [reply .murdock]: Done

**Comment by miri64** [comment]
Addressed comments.

**Comment by kaspar030** [comment]
> do you have any idea how to fix the output ?

Hm, a fix is not trivial. "parse_result.py" in the murdock-scripts repo takes a command like "compile examples/hello-world samr21-xpro" and builds a dict like ``` { "compile" : { "examples/hello-world" : { "board-foo" : ..., "board-bar" : ...} } }```. Also, even if we'd change that assumption to expect another level for the toolchain, the switchover to that other arsing script would be at least messy.

I suggest we encode the toolchain in the board name. Thus, instead of

```compile examples/hello-world samr21-xpro gnu```

we emit

```compile examples/hello-world samr21-xpro:gnu```,

then split again in ```compile()```

**Comment by miri64** [comment]
Like this ad9c08b?

File: .murdock [hunk 3]
```diff
@@ -116,12 +132,12 @@ compile() {
...
 
     # sanity checks
-    [ $# -ne 2 ] && error "$0: compile: invalid parameters (expected \$appdir \$board)"
+    [ $# -ne 2 ] && error "$0: compile: invalid parameters (expected \$appdir \$board:\$toolchain)"
     [ ! -d "$appdir" ] && error "$0: compile: error: application directory \"$appdir\" doesn't exist"
     [ ! -d "boards/$board" ] && error "$0: compile: error: board directory \"boards/$board\" doesn't exist"
 
     # compile
-    CCACHE_BASEDIR="$(pwd)" BOARD=$board RIOT_CI_BUILD=1 \
+    CCACHE_BASEDIR="$(pwd)" BOARD=$board TOOLCHAIN=$toolchain RIOT_CI_BUILD=1 \
```
Comment (kaspar030) [inline .murdock]: You'll need to add the TOOLCHAIN variable to the two make test calls below. Maybe just export TOOLCHAIN?

Comment (kaspar030) [reply .murdock]: Actually it should only be necessary for ```make test-murdock``` below.

Comment (miri64) [reply .murdock]: Why do test *runs* need a toolchain?

Comment (cladmi) [reply .murdock]: I think because they are testing the executable produced by either gnu or llvm and only one could be failing.
Also they would be executed twice with the same name which can cause problems I imagine.

Comment (miri64) [reply .murdock]: Ok. That makes sense. I'll add it.

**Comment by cladmi** [comment]
Can you try adding a 'set -x' at the beginning of the script, it would give bash execution output.

**Comment by miri64** [comment]
Of the `.murdock` script? I'd need to push for that, because the `make murdock-test` target doesn't run on my PC at all (flashfile is set to ""?)?

**Comment by cladmi** [comment]
Yes in the .murdock script and you need to indeed add a debug commit for that.

**Comment by miri64** [comment]
Now Murdock is completely confused :/

**Comment by cladmi** [comment]
I reverted this to fix murdock. I will check what happens and see if I can still debug it from there.

**Comment by cladmi** [comment]
@kaspar030 we need your help again

**Comment by kaspar030** [comment]
I don't know how, but this PR created >184000 murdock jobs and thus exhausted the 1gig of memory the messaging daemon (disque) is configured to use.

**Comment by kaspar030** [comment]
It might be the shell trace (-x) which multiplied the output...

**Comment by kaspar030** [comment]
I found two more problems (and pushed a fixup right here).

**Comment by miri64** [comment]
None of the tests for are executed for `:llvm` though + there is a weird output in the beginning of the log :-/

**Comment by cladmi** [comment]
At least it is not broken anymore and we can start fixing build the :)

**Comment by miri64** [comment]
(Rebased so fixes in master are included)

**Comment by cladmi** [comment]
I created a new label `toolchain` for all the things related to the compilation/libc/linking part.

**Comment by cladmi** [comment]
I currently disabled executing tests until we fix all compilations issues, it will make the build faster.

**Comment by miri64** [comment]
Rebase to include current master and some other PRs.

**Comment by cladmi** [comment]
Found that the test for cflags was not working for clang…
I have a fix I will push.

**Comment by miri64** [comment]
I added a fix for the native issue here directly as well

**Comment by cladmi** [comment]
I also found this to silent the native warning

``` diff
diff --git a/cpu/native/syscalls.c b/cpu/native/syscalls.c
index 3ec3881..be10400 100644
--- a/cpu/native/syscalls.c
+++ b/cpu/native/syscalls.c
@@ -265,6 +265,8 @@ int puts(const char *s)
     return r;
 }
 
+/* Solve 'format string is not a string literal' */
+__attribute__((__format__ (__printf__, 1, 0)))
 char *make_message(const char *format, va_list argp)
 {
     int size = 100;
```

**Comment by miri64** [comment]
> I also found this to silent the native warning

Mhh... maybe your solution is more elegant. Feel free to override mine :-)

**Comment by miri64** [comment]
@vincent-d 

>> Except for one error in can all seem to be in pkgs...
>
> Funny one, this seems to come from LL_PREPEND_ELEM from utlist.h. Not a clue how to fix it though!

@gebart 

> I think the best way is just to ignore the Wtautological-pointer-compare warning for that line.
> See http://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-via-pragmas
> 
> Use something in utlist.h like the example in the clang manual
> 
> ```c
> #pragma clang diagnostic push
> #pragma clang diagnostic ignored "-Wtautological-pointer-compare"
> 
> Blah
> 
> #pragma clang diagnostic pop
> ```

This seems to me rather to indicate that either `clang` is reporting a false positive here (which is possible because it is getting confused by all the pointer magic in that code and in that case the `can` code should actually be surrounded by the pragmas) or that `*list` might have an endless loop.

**Comment by miri64** [comment]
@cladmi can you move 7e761fe878cf023b0c871a55de3ed15d05c3ea97 to a separate PR so it can be faster backported to 2018.07?

**Comment by miri64** [comment]
@jia200x @aabadie https://ci.riot-os.org/RIOT-OS/RIOT/9398/ed2ad6647b863fc15525e439cc288317d7a7d822/output/compile/examples/lorawan/nucleo-f401re:llvm.txt points to an actual bug in LoRAMAC that should be fixed IMHO.

**Comment by miri64** [comment]
@kaspar030 https://ci.riot-os.org/RIOT-OS/RIOT/9398/ed2ad6647b863fc15525e439cc288317d7a7d822/output/compile/examples/javascript/nucleo-f401re:llvm.txt might also cause a bug in jerryscript.

**Comment by cladmi** [comment]
@miri64 I was doing it but was pretty slow :D

**Comment by miri64** [comment]
I can confirm that `tests/pkg_cayenne_lpp` returns different results with `llvm`. @aabadie any idea what's happening?

**Comment by miri64** [comment]
Some weird typedef errors happen for `stdatomic.h` in `core/include/rmurtex.h`. Types are defined, but headers don't provide the type they are based on. @cladmi can you have a look?

**Comment by miri64** [comment]
> I can confirm that tests/pkg_cayenne_lpp returns different results with llvm (on native) than with GCC. @aabadie any idea what's happening?

For some reason some of the value conversions in both `cayenne_lpp_add_temperature()` and `cayenne_lpp_add_accelerometer()` are off by one :-/.

**Comment by jia200x** [comment]
>@jia200x @aabadie https://ci.riot-os.org/RIOT-OS/RIOT/9398/ed2ad6647b863fc15525e439cc288317d7a7d822/output/compile/examples/lorawan/nucleo-f401re:llvm.txt points to an actual bug in LoRAMAC that should be fixed IMHO.

Sure, I'm on it

**Comment by jia200x** [comment]
@miri64 the bug in LoRaMAC is in the pkg, not in the contrib nor include files :(

**Comment by miri64** [comment]
@jia200x can you provide a patch, or better: fix upstream?

**Comment by jia200x** [comment]
> @jia200x can you provide a patch, or better: fix upstream?

Sure! I will check

**Comment by jia200x** [comment]
> @jia200x can you provide a patch, or better: fix upstream?

Seems to be fixed in upstream. However, updating LoRaMAC will require several steps. Is a patch enough for these purposes?

**Comment by cladmi** [comment]
@jia200x Yes a patch is enough for now, it can also be backported to the release

**Comment by jia200x** [comment]
@cladmi check #9626. That should do the trick

**Comment by cladmi** [comment]
Re-running tests since https://github.com/RIOT-OS/RIOT/pull/9599 is merged.

I will add fix commits after.

**Comment by miri64** [comment]
@cladmi just for clarification, will you move your commits to a separate PR later?

**Comment by cladmi** [comment]
@miri64 yes, some of them are already in PRs, I will mention this PR in them to see them in the logs here.

**Comment by cladmi** [comment]
I just `cherry-pick` the commits here to be sure murdock is happy with the changes.

**Comment by miri64** [comment]
Rebased to current master to recheck with it.

**Comment by miri64** [comment]
(reverted 68888b91a4f29288523f21ebb027a064af77ddf4 to proof #9605 makes sense ;-)

**Comment by miri64** [comment]
@jcarrano can you have a look at the LUA issues that this PR reveals. Most of them should be solvable by adapting the `CFLAGS` for `llvm` in `pkg/lua/Makefile`.

**Comment by miri64** [comment]
On the other hand, @cladmi and me can't reproduce lots of the `-Wparentheses-equality` issues on several machines apart from Murdock (local machines with Ubuntu 16.04, 17.10, and 18.04). Can someone with an Arch Linux (@cgundogan, @kaspar030, ...) maybe test `examples/lua_basic` or `tests/lwip` locally? Question is: why does this happen on Murdock, but not on (older?) local machines?

**Comment by miri64** [comment]
(See also https://github.com/RIOT-OS/RIOT/issues/9645)

**Comment by jcarrano** [comment]
@miri64 I just compiled from master, with TOOLCHAIN=llvm, in `native` and `stm32f4discovery`, both `examples/lua_basic` and `examples/lua_REPL`. Everything is working except for #9663 and a small error of mine that I fixed in #9665.

**Comment by jcarrano** [comment]
@miri64 In my Arch machine, `examples/lwip` builds fine with llvm, and lua too.

**Comment by miri64** [comment]
Then we need to find out, why Murdock is behaving differently. @smlng @kaspar030 are the workers set-up and restarted properly? @cladmi mentioned offline, that Docker might update the containers lazily.

**Comment by cladmi** [comment]
Also if it's possible to have an archive of the currently used image. I could diff with my local docker image and try to find the reason.

**Comment by smlng** [comment]
I rebuild the image for the HAW-ICC workers from scratch and restarted all of them, so the image for those containers is up-to-date with riot docker image. However, there was a power flux at the HAW yesterday and now the HAW cloud server have some trouble - but that's another issue.

TL;DR: CI workers should be updated and running.

**Comment by miri64** [comment]
> TL;DR: CI workers should be updated and running.

Mhh, but the `-Wparentheses-equality` issue that we can't produce locally still persists :-(. E.g. https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_udp/native:llvm.txt.

For reference these are the workers that produce different results than we weren't able to reproduce with neither Docker nor our local setup with respective jobs:

* `e5`
  - [tests/lwip/nucleo-f401re:llvm]
* `e52`
  - [tests/lwip/samr21-xpro:llvm]
* `haw-icc`
  - [examples/lua_REPL/nucleo-f401re:llvm]
  - [examples/lua_REPL/samr21-xpro:llvm]
  - [examples/lua_basic/samr21-xpro:llvm]
  - [tests/lwip_sock_ip/nucleo-f401re:llvm]
* `mobi1.inet.haw-hamburg.de`
  - [examples/lua_REPL/native:llvm]
  - [examples/lua_basic/native:llvm]
  - [tests/lwip/native:llvm]
  - [tests/lwip_sock_udp/nucleo-f401re:llvm]
* `mobi3.inet.haw-hamburg.de`
  - [examples/lua_basic/nucleo-f401re:llvm]
  - [tests/lwip_sock_ip/samr21-xpro:llvm]
  - [tests/lwip_sock_udp/samr21-xpro:llvm]
* `mobi30.inet.haw-hamburg.de`
  - [tests/lwip_sock_ip/native:llvm]
  - [tests/lwip_sock_tcp/native:llvm]
* `riot-ci.inet.haw-hamburg.de`
  - [tests/lwip_sock_tcp/nucleo-f401re:llvm]
  - [[tests/lwip_sock_udp/native:llvm]

[examples/lua_REPL/native:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/examples/lua_REPL/native:llvm.txt
[examples/lua_REPL/nucleo-f401re:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/examples/lua_REPL/nucleo-f401re:llvm.txt
[examples/lua_REPL/samr21-xpro:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/examples/lua_REPL/samr21-xpro:llvm.txt
[examples/lua_basic/native:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/examples/lua_basic/native:llvm.txt
[examples/lua_basic/nucleo-f401re:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/examples/lua_basic/nucleo-f401re:llvm.txt
[examples/lua_basic/samr21-xpro:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/examples/lua_basic/samr21-xpro:llvm.txt
[tests/lwip/native:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip/native:llvm.txt
[tests/lwip/nucleo-f401re:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip/nucleo-f401re:llvm.txt
[tests/lwip/samr21-xpro:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip/samr21-xpro:llvm.txt
[tests/lwip_sock_ip/native:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_ip/native:llvm.txt
[tests/lwip_sock_ip/nucleo-f401re:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_ip/nucleo-f401re:llvm.txt
[tests/lwip_sock_ip/samr21-xpro:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_ip/samr21-xpro:llvm.txt
[tests/lwip_sock_tcp/native:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_tcp/native:llvm.txt
[tests/lwip_sock_tcp/nucleo-f401re:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_tcp/nucleo-f401re:llvm.txt
[tests/lwip_sock_tcp/samr21-xpro:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_tcp/samr21-xpro:llvm.txt
[tests/lwip_sock_udp/native:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_udp/native:llvm.txt
[tests/lwip_sock_udp/nucleo-f401re:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_udp/nucleo-f401re:llvm.txt
[tests/lwip_sock_udp/samr21-xpro:llvm]: https://ci.riot-os.org/RIOT-OS/RIOT/9398/6e3557ff490c6655cee77772374b337499f36263/output/compile/tests/lwip_sock_udp/samr21-xpro:llvm.txt

**Comment by miri64** [comment]
Another rebase to include various fixes through-out the repo (PR numbers are referenced within the merge commit).

**Comment by miri64** [comment]
I also removed all the crud from when we tried to integrate this properly ;-)

**Comment by miri64** [comment]
Rebased due to conflicts.

**Comment by miri64** [comment]
Thank you @kaspar030! :-)

**Comment by miri64** [comment]
Cherry-picked 6f78a7f3310f5da918dd6463e5e3d3511a8400a7 from #9700 to get the I²C errors out of the way.

**Comment by miri64** [comment]
Rebased to current master and also included a `nrf5x`-based board to include it into the LLVM build

**Comment by miri64** [comment]
> Rebased to current master and also included a nrf5x-based board to include it into the LLVM build

Maybe [not for now](https://ci.riot-os.org/RIOT-OS/RIOT/9398/4befc6f8f2f7c0fbf6ca66127f816862a164e43a/output/compile/examples/gnrc_minimal/nrf52dk:llvm.txt) =D

**Comment by miri64** [comment]
Ok. There is one "easy" snipe [left in jerryscript](https://ci.riot-os.org/RIOT-OS/RIOT/9398/b30bc6725d6f9ff4307275a26e754dde5904c27d/output/compile/examples/javascript/samr21-xpro:llvm.txt) (just some type confusion). @kaspar030 do you have time to fix that? Everything else is crypto+ASM related... I won't touch that ;-).

**Comment by cladmi** [comment]
I will add some notes on the wiki on how to fix some of the warnings.

**Comment by cladmi** [comment]
I added some notes in the coding conventions regarding correctly fixing warnings:

https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions#common-compilation-warnings

**Comment by miri64** [comment]
Mhhh... Maybe mention in there, that in general non-literal strings as format strings [should be avoided](https://www.owasp.org/index.php/Format_string_attack) and only in those special cases where a `printf`-like function (or [scanf or strftime or strfmon](https://gcc.gnu.org/onlinedocs/gcc-3.1/gcc/Function-Attributes.html)) is used

**Comment by miri64** [comment]
Rebased

**Comment by cladmi** [comment]
Less and less commits remaining :)

**Comment by miri64** [comment]
Rebased and adapted to #9730.

**Comment by miri64** [comment]
Only static tests fail now :dancer:

**Comment by miri64** [comment]
Rebased

**Comment by miri64** [comment]
@RIOT-OS/maintainers feel free to reschedule this one. Currently I'm just trying to snipe issues with LLVM that occur on some random boards.

**Comment by miri64** [comment]
(but please do keep it in the loop!)

**Comment by gebart** [comment]
@miri64 take a quick look at #9362, I think CCACHE_CPP2 will be necessary for all toolchains

**Comment by kaspar030** [comment]
> Since that commit was introduced by @kaspar030, I would prefer if he opens a PR with that commit cherry-picked, but if he's busy I can also do it

I'm fine with @gebart's commit from #9362, it sets the option for all toolchains and has better documentation.

**Comment by miri64** [comment]
Rebased (and squashed), since #9362 was merged.

**Comment by miri64** [comment]
Here's a head-scratcher for someone else to solve: there are some boards, that build *way* bigger for certain applications (or at least they have ROM overflows now) with LLVM/clang ([sometimes even causing ~10kB ROM overflow](https://ci.riot-os.org/RIOT-OS/RIOT/9398/6c8d9c9dec7805be9162e56416ee3c8c66c4d136/output/compile/tests/lwip_sock_tcp/bluepill:llvm.txt)), while with GCC they are linkable (and [supposably fit the ROM](https://ci.riot-os.org/RIOT-OS/RIOT/9398/6c8d9c9dec7805be9162e56416ee3c8c66c4d136/output/compile/tests/lwip_sock_tcp/bluepill:gnu.txt)). I'm not really motivated to investigate this, since all of those boards won't be included into `TEST_BOARDS_LLVM_COMPILE`, but it's strange nevertheless.

**Comment by miri64** [comment]
Rebased

**Comment by miri64** [comment]
I talked about this PR with @cladmi over lunch and we decided to remove the failing architectures from the build list for add them in follow-ups, so we can finally proceed with this PR.

**Comment by miri64** [comment]
Note to self: old state of `.murdock` is in https://github.com/RIOT-OS/RIOT/commit/4df1bcae2c9c4b999f1a7d8abcd3a416944b01b3

**Comment by miri64** [comment]
@kaspar030 can you have another look?

File: .murdock [hunk 4]
```diff
@@ -9,7 +12,7 @@ export DLCACHE_DIR=${DLCACHE_DIR:-~/.dlcache}
 NIGHTLY=${NIGHTLY:-0}
 RUN_TESTS=${RUN_TESTS:-${NIGHTLY}}
 
-DWQ_ENV="-E BOARDS -E APPS -E NIGHTLY -E RUN_TESTS"
+DWQ_ENV="-E BOARDS -E APPS -E NIGHTLY -E RUN_TESTS -E TOOLCHAIN"
```
Comment (miri64) [inline .murdock]: @cladmi and I weren't really sure if this is required

Comment (cladmi) [reply .murdock]: From what I understand now, it should not be necessary.

The compile tasks including toolchain will be listed by https://github.com/RIOT-OS/RIOT/blob/a8c53ec387b2cc8cfb7da5b65311dc2f97d52feb/.murdock#L125-L127

And set from the argument in https://github.com/RIOT-OS/RIOT/blob/a8c53ec387b2cc8cfb7da5b65311dc2f97d52feb/.murdock#L136-L139

Comment (miri64) [reply .murdock]: I removed this change, so let's see how murdock might react.

Comment (miri64) [reply .murdock]: Seems to work regardless :-)

**Comment by miri64** [comment]
Squashed

**Comment by kaspar030** [comment]
I'm happy here. Good job guys!
